samples/javascript/index.js

```diff
@@ -0,0 +1,253 @@
+        if (eventPayload.data.cacheable) {
+            this.cacheableTaskCount++;
+        }
+    }
+
+    complete() {
+        console.log(`Build ${GRADLE_ENTERPRISE_SERVER_URL}/s/${this.buildId} had ${this.cacheableTaskCount} cacheable tasks`)
+    }
+}
+
+// The event handlers to use to process builds.
+const BUILD_EVENT_HANDLERS = [BuildDurationEventsHandler, CacheableTaskCountHandler];
+
+// Code below is a generic utility for interacting with the Export API.
+class BuildProcessor {
+    constructor(gradleEnterpriseServerUrl, maxConcurrentBuildsToProcess, eventHandlerClasses) {
+        this.gradleEnterpriseServerUrl = gradleEnterpriseServerUrl;
+        this.eventHandlerClasses = eventHandlerClasses;
+        this.allHandledEventTypes = this.getAllHandledEventTypes();
+
+        this.pendingBuilds = [];
+        this.numBuildsInProcess = 0;
+        this.maxConcurrentBuildsToProcess = maxConcurrentBuildsToProcess;
+    }
+
+    start(startTime) {
+        const buildStreamUrl = this.createBuildStreamUrl(startTime);
+
+        createServerSideEventStream(buildStreamUrl, {
+            onopen: () => console.log(`Build stream '${buildStreamUrl}' open`),
+            onerror: event => console.error('Build stream error', event),
+            eventListeners: [
+                {
+                    eventName: 'Build',
+                    eventHandler:event => { this.enqueue(JSON.parse(event.data)); }
+                }
+            ],
+            retry: {
+                interval: 6000,
+                maxRetries: 30
+            }
+        });
+    }
+
+    enqueue(build) {
+        this.pendingBuilds.push(build);
+        this.processPendingBuilds();
+    }
+
+    processPendingBuilds() {
+        if (this.pendingBuilds.length > 0 && this.numBuildsInProcess < this.maxConcurrentBuildsToProcess) {
+            this.processBuild(this.pendingBuilds.shift());
+        }
+    }
+
+    createBuildStreamUrl(startTime) {
+        return `${this.gradleEnterpriseServerUrl}/build-export/v1/builds/since/${startTime}?stream`;
+    }
+
+    // Inspect the methods on the handler class to find any event handlers that start with 'on' followed by the event type like 'onBuildStarted'.
+    // Then take the part of the method name after the 'on' to get the event type.
+    getHandledEventTypesForHandlerClass(handlerClass) {
+        return Object.getOwnPropertyNames(handlerClass.prototype)
+            .filter(methodName => methodName.startsWith('on'))
+            .map(methodName => methodName.substring(2));
+    }
+
+    getAllHandledEventTypes() {
+        return new Set(this.eventHandlerClasses.reduce((eventTypes, handlerClass) => eventTypes.concat(this.getHandledEventTypesForHandlerClass(handlerClass)), []));
+    }
+
+    createBuildEventStreamUrl(buildId) {
+        const types = [...this.allHandledEventTypes].join(',');
+        return `${this.gradleEnterpriseServerUrl}/build-export/v1/build/${buildId}/events?eventTypes=${types}`;
```

Comment:
worth reducing the duplication in creating the `baseUrl` for the `export api` as this is done at least twice in this file?